<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Visual Quality Control and Interpretation • pepdiff</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Visual Quality Control and Interpretation">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">pepdiff</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/basic_workflow.html">Getting Started</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Analysis Methods</h6></li>
    <li><a class="dropdown-item" href="../articles/glm_analysis.html">GLM Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/art_analysis.html">ART Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/pairwise_tests.html">Pairwise Tests</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diagnostics</h6></li>
    <li><a class="dropdown-item" href="../articles/checking_fit.html">Checking Model Fit</a></li>
    <li><a class="dropdown-item" href="../articles/diagnostic_plots.html">Diagnostic Plots</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Visual Quality Control and Interpretation</h1>
            
      

      <div class="d-none name"><code>diagnostic_plots.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pepdiff</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="why-visualise">Why visualise?<a class="anchor" aria-label="anchor" href="#why-visualise"></a>
</h2>
<p>Numbers summarise; plots reveal. A p-value tells you something is
significant, but a volcano plot shows you whether that significance is
driven by a few outliers or a robust pattern. Always plot your data
before trusting the statistics.</p>
<p>This vignette shows what “good” and “bad” look like for each
diagnostic plot.</p>
</div>
<div class="section level2">
<h2 id="creating-example-datasets">Creating example datasets<a class="anchor" aria-label="anchor" href="#creating-example-datasets"></a>
</h2>
<p>We’ll create two datasets: one well-behaved, one problematic.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Good data: clean separation, no outliers, balanced</span></span>
<span><span class="co"># Using 10 reps and 3-fold changes for good power</span></span>
<span><span class="va">make_good_data</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">n_peptides</span> <span class="op">&lt;-</span> <span class="fl">40</span></span>
<span>  <span class="va">n_reps</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span></span>
<span>  <span class="va">peptides</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PEP_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%03d"</span>, <span class="fl">1</span><span class="op">:</span><span class="va">n_peptides</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"GENE_"</span>, <span class="va">LETTERS</span><span class="op">[</span><span class="op">(</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">n_peptides</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span> <span class="fl">26</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">diff_peptides</span> <span class="op">&lt;-</span> <span class="va">peptides</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>  <span class="co"># 30% differential</span></span>
<span></span>
<span>  <span class="va">sim_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html" class="external-link">expand.grid</a></span><span class="op">(</span></span>
<span>    peptide <span class="op">=</span> <span class="va">peptides</span>,</span>
<span>    treatment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ctrl"</span>, <span class="st">"trt"</span><span class="op">)</span>,</span>
<span>    bio_rep <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_reps</span>,</span>
<span>    stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>      gene_id <span class="op">=</span> <span class="va">genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">peptide</span>, <span class="va">peptides</span><span class="op">)</span><span class="op">]</span>,</span>
<span>      base <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/GammaDist.html" class="external-link">rgamma</a></span><span class="op">(</span><span class="va">n_peptides</span>, shape <span class="op">=</span> <span class="fl">8</span>, rate <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">n_reps</span><span class="op">)</span>,</span>
<span>      effect <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">peptide</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">diff_peptides</span> <span class="op">&amp;</span> <span class="va">treatment</span> <span class="op">==</span> <span class="st">"trt"</span>,</span>
<span>                      <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.33</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">diff_peptides</span><span class="op">)</span> <span class="op">*</span> <span class="va">n_reps</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                      <span class="fl">1</span><span class="op">)</span>,</span>
<span>      value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/GammaDist.html" class="external-link">rgamma</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, shape <span class="op">=</span> <span class="fl">20</span>, rate <span class="op">=</span> <span class="fl">20</span> <span class="op">/</span> <span class="op">(</span><span class="va">base</span> <span class="op">*</span> <span class="va">effect</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">peptide</span>, <span class="va">gene_id</span>, <span class="va">treatment</span>, <span class="va">bio_rep</span>, <span class="va">value</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">temp_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".csv"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">sim_data</span>, <span class="va">temp_file</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>  <span class="fu"><a href="../reference/read_pepdiff.html">read_pepdiff</a></span><span class="op">(</span><span class="va">temp_file</span>, id <span class="op">=</span> <span class="st">"peptide"</span>, gene <span class="op">=</span> <span class="st">"gene_id"</span>,</span>
<span>               value <span class="op">=</span> <span class="st">"value"</span>, factors <span class="op">=</span> <span class="st">"treatment"</span>, replicate <span class="op">=</span> <span class="st">"bio_rep"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">good_data</span> <span class="op">&lt;-</span> <span class="fu">make_good_data</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Problematic data: outlier sample, batch effect, systematic missingness</span></span>
<span><span class="va">make_bad_data</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">n_peptides</span> <span class="op">&lt;-</span> <span class="fl">40</span></span>
<span>  <span class="va">n_reps</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span></span>
<span>  <span class="va">peptides</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PEP_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%03d"</span>, <span class="fl">1</span><span class="op">:</span><span class="va">n_peptides</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"GENE_"</span>, <span class="va">LETTERS</span><span class="op">[</span><span class="op">(</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">n_peptides</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span> <span class="fl">26</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">sim_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html" class="external-link">expand.grid</a></span><span class="op">(</span></span>
<span>    peptide <span class="op">=</span> <span class="va">peptides</span>,</span>
<span>    treatment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ctrl"</span>, <span class="st">"trt"</span><span class="op">)</span>,</span>
<span>    bio_rep <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_reps</span>,</span>
<span>    stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>      gene_id <span class="op">=</span> <span class="va">genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">peptide</span>, <span class="va">peptides</span><span class="op">)</span><span class="op">]</span>,</span>
<span>      base <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/GammaDist.html" class="external-link">rgamma</a></span><span class="op">(</span><span class="va">n_peptides</span>, shape <span class="op">=</span> <span class="fl">5</span>, rate <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">n_reps</span><span class="op">)</span>,</span>
<span>      <span class="co"># Outlier: one control sample has 3x higher values</span></span>
<span>      batch_effect <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">treatment</span> <span class="op">==</span> <span class="st">"ctrl"</span> <span class="op">&amp;</span> <span class="va">bio_rep</span> <span class="op">==</span> <span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>      value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/GammaDist.html" class="external-link">rgamma</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, shape <span class="op">=</span> <span class="fl">10</span>, rate <span class="op">=</span> <span class="fl">10</span> <span class="op">/</span> <span class="op">(</span><span class="va">base</span> <span class="op">*</span> <span class="va">batch_effect</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">peptide</span>, <span class="va">gene_id</span>, <span class="va">treatment</span>, <span class="va">bio_rep</span>, <span class="va">value</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Add systematic missingness: low-abundance peptides missing in treatment</span></span>
<span>  <span class="va">low_abundance</span> <span class="op">&lt;-</span> <span class="va">peptides</span><span class="op">[</span><span class="fl">31</span><span class="op">:</span><span class="fl">40</span><span class="op">]</span></span>
<span>  <span class="va">sim_data</span> <span class="op">&lt;-</span> <span class="va">sim_data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">peptide</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">low_abundance</span> <span class="op">&amp;</span> <span class="va">treatment</span> <span class="op">==</span> <span class="st">"trt"</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.7</span>,</span>
<span>                          <span class="cn">NA</span>, <span class="va">value</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">temp_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".csv"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">sim_data</span>, <span class="va">temp_file</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>  <span class="fu"><a href="../reference/read_pepdiff.html">read_pepdiff</a></span><span class="op">(</span><span class="va">temp_file</span>, id <span class="op">=</span> <span class="st">"peptide"</span>, gene <span class="op">=</span> <span class="st">"gene_id"</span>,</span>
<span>               value <span class="op">=</span> <span class="st">"value"</span>, factors <span class="op">=</span> <span class="st">"treatment"</span>, replicate <span class="op">=</span> <span class="st">"bio_rep"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">bad_data</span> <span class="op">&lt;-</span> <span class="fu">make_bad_data</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-level-diagnostics">Data-level diagnostics<a class="anchor" aria-label="anchor" href="#data-level-diagnostics"></a>
</h2>
<div class="section level3">
<h3 id="pca-plot">PCA plot<a class="anchor" aria-label="anchor" href="#pca-plot"></a>
</h3>
<p>PCA (Principal Component Analysis) projects your high-dimensional
data onto 2D. Samples that are similar cluster together.</p>
<p><strong>Good PCA:</strong></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_pca_simple.html">plot_pca_simple</a></span><span class="op">(</span><span class="va">good_data</span><span class="op">)</span></span></code></pre></div>
<p><img src="diagnostic_plots_files/figure-html/pca-good-1.png" class="r-plt" alt="" width="672"></p>
<ul>
<li>Replicates from the same condition cluster together</li>
<li>Groups are separated (if there’s a biological effect)</li>
<li>No wild outliers</li>
</ul>
<p><strong>Problematic PCA:</strong></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_pca_simple.html">plot_pca_simple</a></span><span class="op">(</span><span class="va">bad_data</span><span class="op">)</span></span></code></pre></div>
<p><img src="diagnostic_plots_files/figure-html/pca-bad-1.png" class="r-plt" alt="" width="672"></p>
<p>Warning signs: - <strong>Outlier sample</strong>: One point far from
its group suggests a sample quality issue - <strong>No
separation</strong>: If groups overlap completely, your treatment may
have no effect (or the effect is too small to detect) -
<strong>Unexpected clustering</strong>: Samples clustering by batch/run
date instead of treatment suggests a batch effect</p>
<p><strong>What to do:</strong> - Investigate outliers - check sample
prep notes, consider excluding - If batch effects dominate, you may need
to include batch in your model or use batch correction</p>
</div>
<div class="section level3">
<h3 id="distribution-plots">Distribution plots<a class="anchor" aria-label="anchor" href="#distribution-plots"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_distributions_simple.html">plot_distributions_simple</a></span><span class="op">(</span><span class="va">good_data</span><span class="op">)</span></span></code></pre></div>
<p><img src="diagnostic_plots_files/figure-html/dist-good-1.png" class="r-plt" alt="" width="672"></p>
<p><strong>What to look for:</strong> Similar shapes and locations
across samples. Proteomics data is typically right-skewed - that’s
expected.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_distributions_simple.html">plot_distributions_simple</a></span><span class="op">(</span><span class="va">bad_data</span><span class="op">)</span></span></code></pre></div>
<p><img src="diagnostic_plots_files/figure-html/dist-bad-1.png" class="r-plt" alt="" width="672"></p>
<p>Warning signs: - <strong>Shifted distributions</strong>: One sample
systematically higher/lower suggests normalisation issues -
<strong>Different shapes</strong>: One sample much more spread out
suggests technical problems - <strong>Bimodal distributions</strong>:
May indicate a mixture of populations</p>
<p><strong>What to do:</strong> - Check if data was normalised
appropriately - Investigate shifted samples for technical issues -
Consider whether the shifted sample should be excluded</p>
</div>
<div class="section level3">
<h3 id="missingness-plot">Missingness plot<a class="anchor" aria-label="anchor" href="#missingness-plot"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_missingness_simple.html">plot_missingness_simple</a></span><span class="op">(</span><span class="va">good_data</span><span class="op">)</span></span></code></pre></div>
<p><img src="diagnostic_plots_files/figure-html/miss-good-1.png" class="r-plt" alt="" width="672"></p>
<p><strong>Good missingness:</strong> Random scatter, or no missing
values at all. Missing values occur independently of abundance or
condition.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_missingness_simple.html">plot_missingness_simple</a></span><span class="op">(</span><span class="va">bad_data</span><span class="op">)</span></span></code></pre></div>
<p><img src="diagnostic_plots_files/figure-html/miss-bad-1.png" class="r-plt" alt="" width="672"></p>
<p>Warning signs: - <strong>MNAR pattern</strong>: Low-abundance
peptides preferentially missing. This is “missing not at random” - the
value is missing <em>because</em> it’s low (below detection limit). -
<strong>Condition-specific missingness</strong>: Peptides missing only
in treatment (or only in control) may indicate true biological absence,
or may be technical artifacts.</p>
<p><strong>What to do:</strong> - MNAR is common in proteomics and
difficult to handle properly - Peptides with high missingness often
can’t be reliably analysed - pepdiff doesn’t impute - if data is
missing, the peptide may be excluded from analysis - See peppwR for
understanding missingness patterns and power implications</p>
</div>
</div>
<div class="section level2">
<h2 id="results-level-diagnostics">Results-level diagnostics<a class="anchor" aria-label="anchor" href="#results-level-diagnostics"></a>
</h2>
<p>Let’s run analyses on both datasets:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">good_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compare.html">compare</a></span><span class="op">(</span><span class="va">good_data</span>, compare <span class="op">=</span> <span class="st">"treatment"</span>, ref <span class="op">=</span> <span class="st">"ctrl"</span><span class="op">)</span></span>
<span><span class="va">bad_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compare.html">compare</a></span><span class="op">(</span><span class="va">bad_data</span>, compare <span class="op">=</span> <span class="st">"treatment"</span>, ref <span class="op">=</span> <span class="st">"ctrl"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="volcano-plot">Volcano plot<a class="anchor" aria-label="anchor" href="#volcano-plot"></a>
</h3>
<p>The volcano plot shows effect size (fold change) vs statistical
significance. It’s the most informative single plot for differential
abundance results.</p>
<p><strong>Good volcano:</strong></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_volcano_new.html">plot_volcano_new</a></span><span class="op">(</span><span class="va">good_results</span><span class="op">)</span></span></code></pre></div>
<p><img src="diagnostic_plots_files/figure-html/volcano-good-1.png" class="r-plt" alt="" width="672"></p>
<ul>
<li>
<strong>Symmetric spread</strong>: Roughly equal up and down
regulation</li>
<li>
<strong>Significant hits at edges</strong>: Large fold changes with
low p-values</li>
<li>
<strong>Cloud in centre</strong>: Most peptides show small,
non-significant changes</li>
</ul>
<p><strong>Problematic volcano:</strong></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_volcano_new.html">plot_volcano_new</a></span><span class="op">(</span><span class="va">bad_results</span><span class="op">)</span></span></code></pre></div>
<p><img src="diagnostic_plots_files/figure-html/volcano-bad-1.png" class="r-plt" alt="" width="672"></p>
<p>Warning signs: - <strong>Asymmetric</strong>: All significant hits in
one direction may indicate a global shift (normalisation problem) rather
than true differential expression - <strong>Vertical stripe at
FC=1</strong>: Many significant hits with tiny fold changes suggests
p-value inflation - <strong>Everything significant</strong>: Likely a
technical artifact or analysis error</p>
</div>
<div class="section level3">
<h3 id="p-value-histogram">P-value histogram<a class="anchor" aria-label="anchor" href="#p-value-histogram"></a>
</h3>
<p>The p-value histogram reveals whether your statistical assumptions
are met.</p>
<p><strong>Understanding the shapes:</strong></p>
<p>Under the null hypothesis (no true effects), p-values are uniformly
distributed - every value from 0 to 1 is equally likely. When true
effects exist, those peptides get pulled toward 0, creating a spike.</p>
<p><strong>Good p-value histogram:</strong></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_pvalue_histogram.html">plot_pvalue_histogram</a></span><span class="op">(</span><span class="va">good_results</span><span class="op">)</span></span></code></pre></div>
<p><img src="diagnostic_plots_files/figure-html/pval-good-1.png" class="r-plt" alt="" width="672"></p>
<ul>
<li>
<strong>Uniform base + spike near 0</strong>: This is ideal. The
uniform part represents true nulls, the spike represents true
positives.</li>
<li>The height of the spike relative to the uniform part tells you
roughly what fraction of peptides have real effects.</li>
</ul>
<p><strong>Problematic p-value histogram:</strong></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_pvalue_histogram.html">plot_pvalue_histogram</a></span><span class="op">(</span><span class="va">bad_results</span><span class="op">)</span></span></code></pre></div>
<p><img src="diagnostic_plots_files/figure-html/pval-bad-1.png" class="r-plt" alt="" width="672"></p>
<p>Warning signs: - <strong>U-shape</strong> (spikes at both 0 and 1):
P-value inflation. Your test is anti-conservative - it’s giving too many
small AND too many large p-values. Often indicates violated assumptions.
- <strong>Spike at 1 only</strong>: Test is too conservative. May happen
with very small sample sizes. - <strong>Completely uniform</strong>: No
signal at all. Either no true effects, or not enough power to detect
them.</p>
</div>
<div class="section level3">
<h3 id="fold-change-distribution">Fold change distribution<a class="anchor" aria-label="anchor" href="#fold-change-distribution"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_fc_distribution_new.html">plot_fc_distribution_new</a></span><span class="op">(</span><span class="va">good_results</span><span class="op">)</span></span></code></pre></div>
<p><img src="diagnostic_plots_files/figure-html/fc-good-1.png" class="r-plt" alt="" width="672"></p>
<p><strong>Good:</strong> Centred near zero (log2 scale), roughly
symmetric. Most peptides don’t change much.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_fc_distribution_new.html">plot_fc_distribution_new</a></span><span class="op">(</span><span class="va">bad_results</span><span class="op">)</span></span></code></pre></div>
<p><img src="diagnostic_plots_files/figure-html/fc-bad-1.png" class="r-plt" alt="" width="672"></p>
<p>Warning signs: - <strong>Systematic shift</strong>: Distribution
centred away from zero suggests a global offset (normalisation issue) -
<strong>Bimodal</strong>: Two peaks may indicate batch effects or
distinct populations</p>
</div>
</div>
<div class="section level2">
<h2 id="individual-plot-functions">Individual plot functions<a class="anchor" aria-label="anchor" href="#individual-plot-functions"></a>
</h2>
<p>The <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> method gives you a multi-panel grid. For
individual plots:</p>
<p><strong>Data-level:</strong> - <code>plot_pca_simple(data)</code> -
PCA - <code>plot_distributions_simple(data)</code> - Abundance
distributions - <code>plot_missingness_simple(data)</code> - Missingness
patterns</p>
<p><strong>Results-level:</strong> -
<code>plot_volcano_new(results)</code> - Volcano plot -
<code>plot_pvalue_histogram(results)</code> - P-value histogram -
<code>plot_fc_distribution_new(results)</code> - Fold change
distribution</p>
</div>
<div class="section level2">
<h2 id="exporting-for-publication">Exporting for publication<a class="anchor" aria-label="anchor" href="#exporting-for-publication"></a>
</h2>
<p>All plots are ggplot2 objects, so you can customise and save
them:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_volcano_new.html">plot_volcano_new</a></span><span class="op">(</span><span class="va">good_results</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Treatment vs Control"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"volcano.pdf"</span>, <span class="va">p</span>, width <span class="op">=</span> <span class="fl">6</span>, height <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>You can also modify themes, colours, and labels:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_volcano_new.html">plot_volcano_new</a></span><span class="op">(</span><span class="va">good_results</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"My Custom Volcano Plot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="diagnostic_plots_files/figure-html/customise-example-1.png" class="r-plt" alt="" width="672"></p>
</div>
<div class="section level2">
<h2 id="summary-what-to-check">Summary: what to check<a class="anchor" aria-label="anchor" href="#summary-what-to-check"></a>
</h2>
<p><strong>Before analysis:</strong> 1. PCA - do replicates cluster? Any
outliers? 2. Distributions - similar across samples? 3. Missingness -
random or systematic?</p>
<p><strong>After analysis:</strong> 1. Volcano - symmetric? Hits make
sense? 2. P-value histogram - uniform + spike, or something wrong? 3.
Fold changes - centred at zero?</p>
<p>If something looks off, investigate before trusting the statistics.
Plots often reveal problems that summary numbers hide.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Dan MacLean.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
